plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -1375,15 +1376,39 @@ public synchronized void cleanupQuery(ConnectorSession session)
             return;
         }
 
-        commit();
-
-        clearCurrentTransaction();
-        long transactionId = transaction.get().getTransactionId();
-        ScheduledFuture<?> heartbeatTask = transaction.get().getHeartbeatTask();
-        heartbeatTask.cancel(true);
-
-        // Any failure around aborted transactions, etc would be handled by Hive Metastore commit and TrinoException will be thrown
-        delegate.commitTransaction(identity, transactionId);
+        Throwable failure = null;
+        try {
+            commit();
+        }
+        catch (Throwable commitFailure) {
+            failure = commitFailure;
+        }
+        finally {
+            // we need to kill heartbeat thread and commit hive transaction to avoid transaction leak in HMS
+            try {
+                clearCurrentTransaction();
+                long transactionId = transaction.get().getTransactionId();
+                ScheduledFuture<?> heartbeatTask = transaction.get().getHeartbeatTask();
+                heartbeatTask.cancel(true);
+
+                // Any failure around aborted transactions, etc would be handled by Hive Metastore commit and TrinoException will be thrown
+                delegate.commitTransaction(identity, transactionId);
+            }
+            catch (Throwable cleanupFailure) {
+                if (failure != null) {
+                    if (failure != cleanupFailure) {
+                        failure.addSuppressed(cleanupFailure);
+                    }
+                }
+                else {
+                    failure = cleanupFailure;
+                }
+            }
+            if (failure != null) {
+                Throwables.throwIfUnchecked(failure);
+                throw new RuntimeException(failure);
+            }
```

Comment:
You could also extract the cleanup to  a method, this would allow you to prove you're not swallowing Throwables

```java
try {
 commit;
}
catch (Throwable commitFailure) {
  try {
    cleanup;
  } catch (Throwable cleanupFailure) {
    /* if ...*/ commitFailure.addSuppressed(cleanupFailure);
  }
  throw commitFailure; // rethrowing Throwable directly helps the compiler.... and future maintainers
}

// (commit succeeded)
cleanup;
```

Reply:
reworked - looks nicer - thanks